interfaces/policy/policy.go

```diff
@@ -29,7 +29,88 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
-// TODO: InstallCandidate
+// InstallCandidate represents a candidate snap for installation.
+type InstallCandidate struct {
+	Snap            *snap.Info
+	SnapDeclaration *asserts.SnapDeclaration
+	BaseDeclaration *asserts.BaseDeclaration
+}
+
+func (ic *InstallCandidate) checkSlotRule(slot *snap.SlotInfo, rule *asserts.SlotRule, snapRule bool) error {
+	context := ""
+	if snapRule {
+		context = fmt.Sprintf(" for %q snap", ic.SnapDeclaration.SnapName())
+	}
+	if checkSlotInstallationConstraints(slot, rule.DenyInstallation) == nil {
+		return fmt.Errorf("installation denied over slot %q by rule of interface %q%s", slot.Name, slot.Interface, context)
+	}
+	if checkSlotInstallationConstraints(slot, rule.AllowInstallation) != nil {
+		return fmt.Errorf("installation not allowed over slot %q by rule of interface %q%s", slot.Name, slot.Interface, context)
+	}
+	return nil
+}
+
+func (ic *InstallCandidate) checkPlugRule(plug *snap.PlugInfo, rule *asserts.PlugRule, snapRule bool) error {
+	context := ""
+	if snapRule {
+		context = fmt.Sprintf(" for %q snap", ic.SnapDeclaration.SnapName())
+	}
+	if checkPlugInstallationConstraints(plug, rule.DenyInstallation) == nil {
+		return fmt.Errorf("installation denied over plug %q by rule of interface %q%s", plug.Name, plug.Interface, context)
+	}
+	if checkPlugInstallationConstraints(plug, rule.AllowInstallation) != nil {
+		return fmt.Errorf("installation not allowed over plug %q by rule of interface %q%s", plug.Name, plug.Interface, context)
+	}
+	return nil
+}
+
+func (ic *InstallCandidate) checkSlot(slot *snap.SlotInfo) error {
+	iface := slot.Interface
+	if snapDecl := ic.SnapDeclaration; snapDecl != nil {
+		if rule := snapDecl.SlotRule(iface); rule != nil {
+			return ic.checkSlotRule(slot, rule, true)
+		}
+	}
+	if rule := ic.BaseDeclaration.SlotRule(iface); rule != nil {
+		return ic.checkSlotRule(slot, rule, false)
+	}
+	return nil
+}
+
+func (ic *InstallCandidate) checkPlug(plug *snap.PlugInfo) error {
+	iface := plug.Interface
+	if snapDecl := ic.SnapDeclaration; snapDecl != nil {
+		if rule := snapDecl.PlugRule(iface); rule != nil {
+			return ic.checkPlugRule(plug, rule, true)
+		}
+	}
+	if rule := ic.BaseDeclaration.PlugRule(iface); rule != nil {
+		return ic.checkPlugRule(plug, rule, false)
+	}
+	return nil
+}
```

Comment:
I believe this violates this part of the spec: "When the snap says nothing about a particular plug or slot, the base-declaration is consulted. For clarity, when a connection is being established, the final decision is taken by
evaluating the opinion of the snap plug, then the snap slot, then the base plug, and finally the base slot, in that order."

If we use: `checkPlug()` followed by `checkSlot()` then the order is: snap plug, base plug, snap slot, base slot.